File: grafana-backup/dashboard-exporter.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import os
import sys
import json
import time
import shutil
import tarfile
import requests
import subprocess
import click

baseUrl = 'https://monit-grafana.cern.ch/api'

def updatePathEnding(path):
    if not path.endswith("/"):
        path += "/"
    return path
def getGrafanaAuth(fname):
    if not os.path.exists(fname):
        print("File {} does not exist".format(fname))
        sys.exit(1)
    with open(fname, 'r+') as keyFile:
        SECRET_KEY = json.load(keyFile).get('SECRET_KEY')
    headers = {"Authorization": "Bearer %s" % SECRET_KEY}
    return headers


def createBackUpFiles(dashboard, folderTitle, headers):
    allJson = []

    dashboardUid = dashboard.get('uid')
    path = "grafana/dashboards/%s" % folderTitle

    if not os.path.exists(path):
        try:
            os.makedirs(os.path.join('grafana', 'dashboards', folderTitle))
            print("Successfully created the directory %s " % path)
        except OSError:
            print("Creation of the directory %s failed" % path)
    dashboardUrl = '%s/dashboards/uid/%s' % (baseUrl, dashboardUid)
    print('dashboardUrl', dashboardUrl)
    req = requests.get(dashboardUrl, headers=headers)
    dashboardData = req.json().get('dashboard')

    # Replace characters that might cause harm
    titleOfDashboard = dashboardData.get('title').replace(" ", "").replace(".", "_").replace("/", "_")
    uidOfDashboard = dashboardData.get('uid')

    filenameForFolder = '%s/%s-%s.json' % (path, titleOfDashboard, uidOfDashboard)
    print('filenameForFolder', filenameForFolder)
    with open(filenameForFolder, "w+") as jsonFile:
        json.dump(dashboardData, jsonFile)
    print('*' * 10 + '\n')
    allJson.append(dashboardData)

    filenameForAllJson = '%s/all.json' % path
    print('filenameForAllJson', filenameForAllJson)
    with open(filenameForAllJson, "w+") as jsonFile:
        json.dump(allJson, jsonFile)


def searchFoldersFromGrafana(headers):
    foldersUrl = 'https://monit-grafana.cern.ch/api/search?folderIds=0&orgId=11'

    req = requests.get(foldersUrl, headers=headers)
    if req.status_code != 200:
        print("Request {}, failed with reason: {}".format(foldersUrl, req.reason))
        sys.exit(1)
    foldersJson = req.json()

    for folder in foldersJson:
        folderId = folder.get('id')
        folderTitle = folder.get('title')

        if folderTitle not in ['Production', 'Development', 'Playground', 'Backup']:
            folderTitle = 'General'
            dashboard = folder
            createBackUpFiles(dashboard, folderTitle, headers)

        else:
            dashboardQuery = 'search?folderIds=%s&orgId=11&query=' % folderId
            dashboardQueryUrl = '%s/%s' % (baseUrl, dashboardQuery)

            print('individualFolderUrl', dashboardQueryUrl)

            req = requests.get(dashboardQueryUrl, headers=headers)
            folderData = req.json()

            pathToDb = os.path.abspath("grafana")
            if not os.path.exists(pathToDb):
                os.makedirs(pathToDb)

            filenameForFolder = 'grafana/%s-%s.json' % (folderId, folderTitle)
            with open(filenameForFolder, "w+") as jsonFile:
                json.dump(folderData, jsonFile)

            dashboardsAmount = len(folderData)
            for index, dashboard in enumerate(folderData):
                folderTitle = dashboard.get('folderTitle')
                print('Index/Amount: %s/%s' % (index, dashboardsAmount))

                createBackUpFiles(dashboard, folderTitle, headers)


def createTar(path, tar_name):
    with tarfile.open(tar_name, "w:gz") as tar_handle:
        for root, _, files in os.walk(path):
            for file in files:
                tar_handle.add(os.path.join(root, file))


def removeTempFiles(path):
    for root, dirs, files in os.walk(path):
        for f in files:
            os.unlink(os.path.join(root, f))
        for d in dirs:
            shutil.rmtree(os.path.join(root, d))
    if os.path.exists(path):
        os.rmdir(path)


def get_date():
    gmt = time.gmtime()
    year = str(gmt.tm_year)
    mon = gmt.tm_mon
    if len(str(mon)) == 1:
        mon = '0{}'.format(mon)
    else:
        mon = str(mon)
    day = gmt.tm_mday
    if len(str(day)) == 1:
        day = '0{}'.format(day)
    else:
        day = str(day)
    return day, mon, year


def copyToHDFS(archive, hdfs_path):
    day, mon, year = get_date()
    hdfs_path = updatePathEnding(hdfs_path)
    path = f'{hdfs_path}{year}/{mon}/{day}'
    print("Copy backup to {}".format(path))
    cmd = 'hadoop fs -mkdir -p {}'.format(path)
    output = subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True)
    if output:
        print(output)
    cmd = 'hadoop fs -put {} {}'.format(archive, path)
    print('Execute: {}'.format(cmd))
    output = subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True)
    if output:
        print(output)
    cmd = 'hadoop fs -ls {}'.format(path)
    output = subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True)
    print("New backup: {}".format(output))


def copyToFileSystem(archive, base_dir):
    day, mon, year = get_date()
    base_dir = updatePathEnding(base_dir)
    path = f'{base_dir}{year}/{mon}/{day}'
    print("Copy backup to {}".format(path))
    cmd = 'mkdir -p {}'.format(path)
    output = subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True)
    if output:
        print(output)
    cmd = 'cp {} {}'.format(archive, path)
    print('Execute: {}'.format(cmd))
    output = subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True)
    if output:
        print(output)
    cmd = 'ls {}'.format(path)
    output = subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True)
    print("New backup: {}".format(output))


@click.command()
@click.option("--token", "fname", required=True, help="Grafana token location JSON file")
@click.option("--hdfs-path", required=True, help="Path for Grafana backup in HDFS")
@click.option("--filesystem-path", required=True, help="Path for Grafana backup in EOS")
def main(fname, hdfs_path, filesystem_path):
    click.echo(f"Input Arguments: token:{fname}, hdfs_path:{hdfs_path}, filesystem_path:{filesystem_path}")
    headers = getGrafanaAuth(fname)
    searchFoldersFromGrafana(headers)
    createTar('./grafana', 'grafanaBackup.tar.gz')
    copyToHDFS('grafanaBackup.tar.gz', hdfs_path)
    copyToFileSystem('grafanaBackup.tar.gz', filesystem_path)
    removeTempFiles('./grafana/')


if __name__ == '__main__':
    main()

File: grafana-backup/dashboard-exporter.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/urfave/cli/v2"
	"archive/tar" // Import the "tar" package for tar file operations
)


const baseUrl = "https://monit-grafana.cern.ch/api"

type Dashboard struct {
	Uid   string `json:"uid"`
	Title string `json:"title"`
}

func updatePathEnding(path string) string {
	if !strings.HasSuffix(path, "/") {
		path += "/"
	}
	return path
}

func getGrafanaAuth(fname string) (string, error) {
	if _, err := os.Stat(fname); os.IsNotExist(err) {
		return "", fmt.Errorf("File %s does not exist", fname)
	}

	file, err := os.Open(fname)
	if err != nil {
		return "", err
	}
	defer file.Close()

	var data map[string]interface{}
	err = json.NewDecoder(file).Decode(&data)
	if err != nil {
		return "", err
	}

	secretKey, ok := data["SECRET_KEY"].(string)
	if !ok {
		return "", fmt.Errorf("Failed to get SECRET_KEY from file")
	}

	headers := fmt.Sprintf("Bearer %s", secretKey)
	return headers, nil
}

func createBackUpFiles(dashboard Dashboard, folderTitle, headers string) {
	// Function to create backup files
}

func searchFoldersFromGrafana(headers string) {
	// Function to search folders and perform backup operations
}

func createTar(path, tarName string) error {
	with, err := os.Create(tarName)
	if err != nil {
		return err
	}
	defer with.Close()

	tw := tar.NewWriter(with)
	defer tw.Close()

	files, err := os.ReadDir(path)
	if err != nil {
		return err
	}

	for _, file := range files {
		if err := addFileToTar(tw, filepath.Join(path, file.Name()), ""); err != nil {
			return err
		}
	}

	return nil
}

func addFileToTar(tw *tar.Writer, path, prefix string) error {
	file, err := os.Open(path)
	if err != nil {
		return err
	}
	defer file.Close()

	info, err := file.Stat()
	if err != nil {
		return err
	}

	header, err := tar.FileInfoHeader(info, "")
	if err != nil {
		return err
	}
	header.Name = filepath.Join(prefix, filepath.Base(path))

	if err := tw.WriteHeader(header); err != nil {
		return err
	}

	if _, err := io.Copy(tw, file); err != nil {
		return err
	}

	return nil
}

func removeTempFiles(path string) error {
	err := os.RemoveAll(path)
	if err != nil {
		return err
	}
	return nil
}

func getDate() (string, string, string) {
	gmt := time.Now().UTC()
	year := fmt.Sprintf("%04d", gmt.Year())
	mon := fmt.Sprintf("%02d", int(gmt.Month()))
	day := fmt.Sprintf("%02d", gmt.Day())
	return day, mon, year
}

func copyToHDFS(archive, hdfsPath string) error {
	day, mon, year := getDate()
	hdfsPath = updatePathEnding(hdfsPath)
	path := fmt.Sprintf("%s%s/%s/%s", hdfsPath, year, mon, day)

	if _, err := exec.Command("hadoop", "fs", "-mkdir", "-p", path).CombinedOutput(); err != nil {
		return err
	}

	copyCmd := exec.Command("hadoop", "fs", "-put", archive, path)
	copyCmd.Stdout = os.Stdout
	copyCmd.Stderr = os.Stderr

	if err := copyCmd.Run(); err != nil {
		return err
	}

	listCmd := exec.Command("hadoop", "fs", "-ls", path)
	listCmd.Stdout = os.Stdout
	listCmd.Stderr = os.Stderr
	if err := listCmd.Run(); err != nil {
		return err
	}

	return nil
}

func copyToFileSystem(archive, baseDir string) error {
	day, mon, year := getDate()
	baseDir = updatePathEnding(baseDir)
	path := fmt.Sprintf("%s%s/%s/%s", baseDir, year, mon, day)

	if err := os.MkdirAll(path, os.ModePerm); err != nil {
		return err
	}

	copyCmd := exec.Command("cp", archive, path)
	copyCmd.Stdout = os.Stdout
	copyCmd.Stderr = os.Stderr

	if err := copyCmd.Run(); err != nil {
		return err
	}

	listCmd := exec.Command("ls", path)
	listCmd.Stdout = os.Stdout
	listCmd.Stderr = os.Stderr
	if err := listCmd.Run(); err != nil {
		return err
	}

	return nil
}

func main() {
	app := &cli.App{
		Name:  "Grafana Backup",
		Usage: "A tool to backup Grafana dashboards",
		Flags: []cli.Flag{
			&cli.StringFlag{
				Name:     "token",
				Usage:    "Grafana token location JSON file",
				Required: true,
			},
			&cli.StringFlag{
				Name:     "filesystem-path",
				Usage:    "Path for Grafana backup in EOS",
				Required: true,
			},
		},
		Action: func(c *cli.Context) error {
			tokenFile := c.String("token")
			filesystemPath := c.String("filesystem-path")

			headers, err := getGrafanaAuth(tokenFile)
			if err != nil {
				return err
			}

			searchFoldersFromGrafana(headers)

			if err := createTar("./grafana", "grafanaBackup.tar.gz"); err != nil {
				return err
			}

			if err := copyToFileSystem("grafanaBackup.tar.gz", filesystemPath); err != nil {
				return err
			}

			if err := removeTempFiles("./grafana/"); err != nil {
				return err
			}

			return nil
		},
	}

	err := app.Run(os.Args)
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}
